mafia

```diff
@@ -0,0 +1,77 @@
+#!/bin/sh -eu
+
+fetch_latest () {
+  if [ -z ${MAFIA_TEST_MODE+x} ]; then
+    TZ=$(date +"%T")
+    curl --silent "https://raw.githubusercontent.com/ambiata/mafia/master/script/mafia?$TZ"
+  else
+    cat ../script/mafia
+  fi
+}
+
+run_upgrade () {
+  echo "Checking for a new version of mafia ..."
+  fetch_latest > /tmp/mafia
+
+  COMMIT_VERSION=$(git ls-remote https://github.com/ambiata/mafia | grep refs/heads/master | cut -f 1)
+  echo "# Version: $COMMIT_VERSION" >> /tmp/mafia
+
+  if ! cmp ./mafia /tmp/mafia >/dev/null 2>&1; then
+    echo "New version found and upgraded. You can now commit it to your git repo."
+    mv /tmp/mafia ./mafia
+    chmod +x ./mafia
+  else
+    echo "You have latest mafia script"
+  fi
+}
+
+exec_mafia () {
+  MAFIA_VERSION=$(awk '/^# Version: / { print $3; exit 0; }' $0)
+
+  if [ "x$MAFIA_VERSION" == "x" ]; then
+    # If we can't find the mafia version, then we need to upgrade the script.
+    run_upgrade
+  else
+    MAFIA_PATH=$HOME/.ambiata/mafia/bin/mafia-$MAFIA_VERSION
+
+    [ -f "$MAFIA_PATH" ] || {
+      # Create a temporary directory which will be deleted when the script
+      # terminates. Unfortunately `mktemp` doesn't behave the same on
+      # Linux and OS/X so we need to try two different approaches.
+      MAFIA_TEMP=$(mktemp -d 2>/dev/null || mktemp -d -t 'exec_mafia')
+      trap "rm -rf \"$MAFIA_TEMP\"" EXIT
+
+      echo "Building mafia-$MAFIA_VERSION in $MAFIA_TEMP"
+
+      ( cd "$MAFIA_TEMP"
+
+        git clone https://github.com/ambiata/mafia
+        cd mafia
+
+        git reset --hard $MAFIA_VERSION
+
+        bin/bootstrap ) || exit $?
+
+      mkdir -p $(dirname $MAFIA_PATH)
+      cp "$MAFIA_TEMP/mafia/.cabal-sandbox/bin/mafia" "$MAFIA_PATH"
```

Comment:
Ug, @jystic Sorry, I missed this, there is a race condition here that ci will probably hit after this gets rolled out to more versions. This should be something more like:
```
cp "$MAFIA_TEMP/mafia/.cabal-sandbox/bin/mafia" "$MAFIA_PATH.tmp"
mv "$MAFIA_PATH.tmp" "$MAFIA_PATH"
```
Otherwise a build will (potentially) end up with half an executable.